lib/trino-memory-context/src/main/java/io/trino/memory/context/MemoryAllocationValidator.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.memory.context;
+
+public interface MemoryAllocationValidator
+{
+    MemoryAllocationValidator NO_MEMORY_VALIDATION = new MemoryAllocationValidator() {
+        @Override
+        public void reserveMemory(String allocationTag, long delta) {}
+
+        @Override
+        public boolean tryReserveMemory(String allocationTag, long delta)
+        {
+            return true;
+        }
+    };
+
+    /**
+     * Check if memory can be reserved. Account for reserved memory if reservation is possible and return true. Throw exception otherwise.
+     */
+    void reserveMemory(String allocationTag, long delta);
+
+    /**
+     * Check if memory can be reserved. Account for reserved memory if reservation is possible and return true. Return false otherwise.
```

Comment:
Do we need two functions. Take a look at `io.trino.memory.QueryContext#enforceUserMemoryLimit`. It seems just a single function is needed to enforce memory limits.

Reply:
Two methods there so we do not need to do exception driven logic in `ChildAggregatedMemoryContext.tryUpdateBytes` where `MemoryAllocationValidator.tryReserveMemory` is used. 
If I just had `reserveMemory` I would need to `try/catch` there. It is an option, but I think it is not ellegant.